evap/staff/views.py

```diff
@@ -144,159 +149,192 @@ def __init__(self):
+
+    @staticmethod
+    def apply(request, evaluations, email_template=None):
+        for evaluation in evaluations:
+            evaluation.publish()
+            evaluation.save()
+        messages.success(request, ungettext("Successfully published {} evaluation.",
+            "Successfully published {} evaluations.", len(evaluations)).format(len(evaluations)))
+        if email_template:
+            send_publish_notifications(evaluations, email_template)
+
+
+EVALUATION_OPERATIONS = {
+        'new': RevertToNewOperation,
+        'prepared': RevertToPreparedOperation,
+        'in_evaluation': StartEvaluationOperation,
+        'reviewed': RevertToReviewedOperation,
+        'published': PublishOperation,
+}
+
+
 @manager_required
 def semester_evaluation_operation(request, semester_id):
     semester = get_object_or_404(Semester, id=semester_id)
     if semester.participations_are_archived:
         raise PermissionDenied
 
     target_state = request.GET.get('target_state')
-    if target_state not in ['new', 'prepared', 'in_evaluation', 'reviewed', 'published']:
+    if target_state not in EVALUATION_OPERATIONS.keys():
         raise SuspiciousOperation("Unknown target state: " + target_state)
 
     evaluation_ids = (request.GET if request.method == 'GET' else request.POST).getlist('evaluation')
-    evaluations = Evaluation.objects.filter(id__in=evaluation_ids).annotate(
-        midterm_grade_documents_count=Count("course__grade_documents", filter=Q(course__grade_documents__type=GradeDocument.MIDTERM_GRADES), distinct=True),
-        final_grade_documents_count=Count("course__grade_documents", filter=Q(course__grade_documents__type=GradeDocument.FINAL_GRADES), distinct=True)
-    )
+    evaluations = annotate_evaluations_with_grade_document_counts(Evaluation.objects.filter(id__in=evaluation_ids))
+    operation = EVALUATION_OPERATIONS[target_state]
 
     if request.method == 'POST':
-        template = None
+        email_template = None
         if request.POST.get('send_email') == 'on':
-            template = EmailTemplate(subject=request.POST['email_subject'], body=request.POST['email_body'])
-
-        if target_state == 'new':
-            helper_semester_evaluation_operation_revert(request, evaluations)
-        elif target_state == 'prepared':
-            helper_semester_evaluation_operation_prepare(request, evaluations, template)
-        elif target_state == 'in_evaluation':
-            helper_semester_evaluation_operation_start(request, evaluations, template)
-        elif target_state == 'reviewed':
-            helper_semester_evaluation_operation_unpublish(request, evaluations)
-        elif target_state == 'published':
-            helper_semester_evaluation_operation_publish(request, evaluations, template)
-
+            email_template = EmailTemplate(subject=request.POST['email_subject'], body=request.POST['email_body'])
+        operation.apply(request, evaluations, email_template)
```

Comment:
remember this needs to be updated when #1317 is merged

Reply:
this will be merged first

Reply:
We agreed to change 1317 afterwards (as it makes more sense for me to do the merging)